client/src/app/core/ui-services/tree.service.ts

```diff
@@ -56,6 +77,169 @@ export class TreeService {
+        }
+
+        // Sort the parents
+        parents.sort((a, b) => this.getAttributeAsNumber(a, weightKey) - this.getAttributeAsNumber(b, weightKey));
+
+        // Iterate through the parents to see, who has children
+        let i = 0;
+        for (const value of parents) {
+            const tmpArray = this.buildTreeHelper(0, value, mapChildren);
+            osTreeNodes.push(...tmpArray);
+            i += tmpArray.length;
+        }
+
+        for (let j = 0; j < osTreeNodes.length; ++j) {
+            osTreeNodes[j].position = j;
+        }
+
+        return osTreeNodes;
+    }
+
+    /**
+     * Function that should help to build a flat tree from the given items.
+     * If the map contains the id of the given item as key, than it iterates through the children and calls this function recursively to get further children.
+     *
+     * @param position is the position of the item in the list.
+     * @param level is the level of the item related to its parent.
+     * @param parent is the parent of the given item.
+     * @param mapChildren is a map containing the parentId as number as key and the related child nodes as array as value.
+     *
+     * @returns an array that contains the parent inclusive its children.
+     */
+    private buildTreeHelper<T extends Identifiable & Displayable>(
+        level: number,
+        parent: T,
+        mapChildren: Map<Number, T[]>
+    ): OSFlatNode[] {
+        const children = mapChildren.get(parent.id);
+
+        // build the parent node
+        const tmpArray: OSFlatNode[] = [
+            {
+                id: parent.id,
+                name: parent.getTitle(),
+                level: level,
+                expandable: !!children,
+                isExpanded: !!children
+            }
+        ];
+
+        // if there are children, check if they have children, too
+        if (children) {
+            for (let i = 0; i < children.length; ++i) {
+                const value = children[i];
+                tmpArray.push(...this.buildTreeHelper(level + 1, value, mapChildren));
+            }
+        }
+        return tmpArray;
+    }
+
+    /**
+     * Function, that returns a node containing information like id, name and children.
+     * Children only, if available.
+     *
+     * @param node The node which is converted.
+     * @param nodes The array with all nodes to convert.
+     * @param length The number of converted nodes related to the parent node.
+     *
+     * @returns `OSTreeNodeWithOutItem`
+     */
+    private getNodeWithChildren(
+        node: OSFlatNode,
+        nodes: OSFlatNode[],
+        parentId: number,
+        length: number
+    ): { tree: Array<OSTreeSortEvent>; length: number } {
+        const children = [];
+        // Begins at the position of the node in the array
+        // Ends if the next node has the same or higher level than the given node.
+        for (let i = node.position + 1; !!nodes[i] && nodes[i].level >= node.level + 1; ++i) {
```

Comment:
This is the only place node.position is used. To keep things easier, you can pass the current index as a parameter to this function and remove `position` everywhere. Is it used in the sorted-tree-component?

Reply:
In fact, that position is nothing else as the index of the node, maybe there is a possibility to always use the index. It would reduce redundancy.

Reply:
As discussed the position will remain in the `interface FlatNode`, because the only way to get the index of the node is to search for it, but this is a huge more effort.